ENH new svd_solver='covariance_eigh' for PCA

ogrisel opened the pull request:
Implements the new solver proposed in #27483.

I also took the opportunity of this PR to strengthen the solver equivalence tests.

## Benchmark

### Laptop CPU with numpy on macOS (with Apple Accelerate BLAS)

```python
>>> import numpy as np
>>> from sklearn.datasets import make_low_rank_matrix
>>> X = make_low_rank_matrix(n_samples=int(1e6), n_features=100).astype(np.float32)
>>> from sklearn.decomposition import PCA
>>> %time PCA(n_components=5, svd_solver="randomized").fit_transform(X)[-1]
CPU times: user 1.67 s, sys: 186 ms, total: 1.85 s
Wall time: 1.1 s
array([ 0.00055511, -0.00047962,  0.0009717 , -0.00128793, -0.00048758],
      dtype=float32
>>> %time PCA(n_components=5, svd_solver="full").fit_transform(X)[-1]
CPU times: user 11.8 s, sys: 298 ms, total: 12.1 s
Wall time: 3.56 s
array([ 0.00055506, -0.00047963,  0.00097174, -0.00128779, -0.00048765],
      dtype=float32)
>>> %time PCA(n_components=5, svd_solver="covariance_eigh").fit_transform(X)[-1]
CPU times: user 121 ms, sys: 3.3 ms, total: 124 ms
Wall time: 123 ms
array([ 0.00055505, -0.00047963,  0.00097174, -0.0012878 , -0.00048764],
      dtype=float32)
```

so around a 10x speed-up.

Switching to OpenBLAS makes it 2x or so slower than Apple Accelerate but still very good.

Note: to switch the BLAS implementation with conda-forge, there is no need to rebuild scikit-learn or scipy. Just the following commands:

```
mamba install "libblas=*=*accelerate"
```
or
```
mamba install "libblas=*=*openblas"
```

### PyTorch via Array API with MPS device on Apple M1

```python
>>> import numpy as np
>>> from sklearn.datasets import make_low_rank_matrix
>>> X = make_low_rank_matrix(n_samples=int(1e6), n_features=100).astype(np.float32)
>>> import torch
>>> X_torch_mps = torch.tensor(X).to("mps")
>>> import sklearn
>>> sklearn.set_config(array_api_dispatch=True)
>>> from sklearn.decomposition import PCA
>>> %time PCA(n_components=5, svd_solver="randomized", power_iteration_normalizer="QR").fit_transform(X_torch_mps)[-1]
/Users/ogrisel/code/scikit-learn/sklearn/utils/extmath.py:319: UserWarning: The operator 'aten::linalg_qr.out' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/miniforge3/conda-bld/pytorch-recipe_1690825982320/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  Q, _ = normalizer(A @ Q)
CPU times: user 3.48 s, sys: 474 ms, total: 3.96 s
Wall time: 3.02 s
tensor([ 0.0003, -0.0009,  0.0004, -0.0002, -0.0002], device='mps:0')
>>> %time PCA(n_components=5, svd_solver="full").fit_transform(X_torch_mps)[-1]
CPU times: user 11.4 s, sys: 483 ms, total: 11.9 s
Wall time: 3.83 s
tensor([ 0.0003, -0.0009,  0.0004, -0.0002, -0.0002], device='mps:0'
>>> %time PCA(n_components=5, svd_solver="covariance_eigh").fit_transform(X_torch_mps)[-1]
CPU times: user 11.6 ms, sys: 69.7 ms, total: 81.4 ms
Wall time: 368 ms
tensor([ 0.0003, -0.0009,  0.0004, -0.0002, -0.0002], device='mps:0')
```

As the warning said, the MPS backend for PyTorch still lacks `torch.linalg.*` kernels and fallbacks to the CPU which makes the overall performance a bit slower than just using numpy on CPU. I can also try directly using PyTorch on the CPU (without the MPS backend):

```python
>>> X_torch_cpu = torch.tensor(X)
>>> %time PCA(n_components=5, svd_solver="randomized", power_iteration_normalizer="QR").fit_transform(X_torch_cpu)[-1]
CPU times: user 3.12 s, sys: 692 ms, total: 3.81 s
Wall time: 1.44 s
tensor([ 0.0003, -0.0009,  0.0004, -0.0002, -0.0002])
>>> %time PCA(n_components=5, svd_solver="full").fit_transform(X_torch_cpu)[-1]
CPU times: user 11.3 s, sys: 535 ms, total: 11.8 s
Wall time: 3.23 s
tensor([ 0.0003, -0.0009,  0.0004, -0.0002, -0.0002])
>>> %time PCA(n_components=5, svd_solver="covariance_eigh").fit_transform(X_torch_cpu)[-1]
CPU times: user 198 ms, sys: 18.3 ms, total: 217 ms
Wall time: 112 ms
tensor([ 0.0003, -0.0009,  0.0004, -0.0002, -0.0002])
```

So we get similar results as with numpy on the same host.

### Intel Xeon CPU (44 physical cores from 2018 or so) with numpy on Linux (with MKL BLAS)

I get similar numbers (slightly slower) as on my macOS laptop for the new solver. I inspected the use of threads with htop (on a 10x larger dataset) and I could see that MKL (and OpenBLAS) decide to only use 1 thread for the matrix matrix multiplication (to compute the covariance). This is probably because the matrix is very skinny and concurrent updates by several threads would be detrimental. As a result, using a multi-core machine is not very beneficial for this solver on this kind of data (but it's still more 5x faster than using the randomized solver that actually uses many cores on htop, but not 100% of the time).

Switching to OpenBLAS makes it 2x or so slower than MKL but still very good.

Anyways, the performance (with MKL) seems to be competitive with the [scikit-learn-intelex benchmark results](https://mbatoul.github.io/sklearn_benchmarks/results/github_ci/master/20220322T183227/scikit_learn_intelex_vs_scikit_learn.html#speedup-barplots-pca) we collected a while ago. Note that I cannot reproduce those values with the latest version of scikit-learn-intelex because it seems to use the scikit-learn "randomized" solver instead of their own in the latest version.

### PyTorch via Array API with cuda (NVIDIA V100 GPU)

```python
>>> import numpy as np
>>> from sklearn.datasets import make_low_rank_matrix
>>> X = make_low_rank_matrix(n_samples=int(1e6), n_features=100).astype(np.float32)
>>> import torch
>>> X_torch_cuda = torch.tensor(X).to("cuda")
>>> import sklearn
>>> sklearn.set_config(array_api_dispatch=True)
>>> from sklearn.decomposition import PCA
>>> %time PCA(n_components=5, svd_solver="randomized", power_iteration_normalizer="QR").fit_transform(X_torch_cuda)[-1]
CPU times: user 63.8 ms, sys: 15.5 ms, total: 79.3 ms
Wall time: 78.3 ms
tensor([ 5.4981e-06,  7.2371e-04,  5.3485e-04, -1.3649e-03, -1.4326e-03],
       device='cuda:0')
>>> %time PCA(n_components=5, svd_solver="full").fit_transform(X_torch_cuda)[-1]
CPU times: user 81.4 ms, sys: 19.7 ms, total: 101 ms
Wall time: 99.8 ms
tensor([ 5.5376e-06,  7.2371e-04,  5.3483e-04, -1.3650e-03, -1.4326e-03],
       device='cuda:0')
>>> %time PCA(n_components=5, svd_solver="covariance_eigh").fit_transform(X_torch_cuda)[-1]
CPU times: user 7.67 ms, sys: 362 µs, total: 8.04 ms
Wall time: 7.21 ms
tensor([ 5.4768e-06,  7.2370e-04,  5.3485e-04, -1.3650e-03, -1.4326e-03],
       device='cuda:0')
```

So here again we observe a 10x speed-up with the new solver.

We can compare to NVIDIA's cuML from RAPIDS 23.08 on the same data either provided as CPU allocated numpy array or GPU allocated CuPy:

```python
>>> import numpy as np
>>> from sklearn.datasets import make_low_rank_matrix
>>> X = make_low_rank_matrix(n_samples=int(1e6), n_features=100).astype(np.float32)
>>> from cuml.decomposition import PCA
>>> %time PCA(n_components=5).fit_transform(X)[-1]
CPU times: user 177 ms, sys: 42.5 ms, total: 220 ms
Wall time: 224 ms
array([-0.00171027,  0.00036515, -0.00089102,  0.00108428, -0.00054887],
      dtype=float32)
>>> import cupy
>>> X_cupy = cupy.asarray(X)
>>> %time PCA(n_components=5).fit_transform(X_cupy)[-1]
CPU times: user 25.3 ms, sys: 11.5 ms, total: 36.8 ms
Wall time: 36.8 ms
array([-0.00171027,  0.00036515, -0.00089102,  0.00108428, -0.00054887],
      dtype=float32)
```

So when the data is pre-allocated on the cuda device, our PyTorch + cuda implementation is 5x faster than cuML's!

### PyTorch via Array API on Intel Max GPU

The performance is not good at all, but I suspect that the Intel extension for PyTorch is not ready yet. I opened an issue to track progress upstream:

- https://github.com/intel/intel-extension-for-pytorch/issues/439

## TODO:

- [x] add a changelog entry,
- [x] run some benchmarks,
- [x] check that Array API compliance tests works on cuda, including with the new solver
- [x] compare the new solver to cuML on GPU
- [x] fix remaining CI failures,
- [x] improve the test by also checking consistent output for `transform` and `fit_transform`,
- [x] add a new commit to use the new solver when it makes sense by updating the dimension dependent benchmark scripts (to be discussed next),
- [x] compare the new solver to scikit-learn-intelex on CPU

Note that I had to change the way we use `svd_flip` in this estimator to be able to use get the same signs for the components of the new solver. I think it's better for scikit-learn to be self-consistent (checked in the solver equivalence test) than consistent across versions.

ogrisel committed:
ENH new svd_solver='covariance_eigh' for PCA

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel committed:
Add changelog entry

ogrisel committed:
Change test parametrization to workaround pytest xdist bug

ogrisel committed:
Avoid raising warning when dtype is None

ogrisel commented:
I get a pytest xdist bug on the CI:

```
_____________________________ ERROR collecting gw0 _____________________________
Different tests were collected between gw1 and gw0. The difference is:
--- gw1

+++ gw0

@@ -4099,12 +4099,12 @@

 decomposition/tests/test_pca.py::test_whitening[randomized-False]
 decomposition/tests/test_pca.py::test_whitening[auto-True]
 decomposition/tests/test_pca.py::test_whitening[auto-False]
+decomposition/tests/test_pca.py::test_pca_solver_equivalence[42-arpack-tall]
+decomposition/tests/test_pca.py::test_pca_solver_equivalence[42-arpack-wide]
 decomposition/tests/test_pca.py::test_pca_solver_equivalence[42-randomized-tall]
 decomposition/tests/test_pca.py::test_pca_solver_equivalence[42-randomized-wide]
 decomposition/tests/test_pca.py::test_pca_solver_equivalence[42-covariance_eigh-tall]
 decomposition/tests/test_pca.py::test_pca_solver_equivalence[42-covariance_eigh-wide]
-decomposition/tests/test_pca.py::test_pca_solver_equivalence[42-arpack-tall]
-decomposition/tests/test_pca.py::test_pca_solver_equivalence[42-arpack-wide]
 decomposition/tests/test_pca.py::test_pca_explained_variance_empirical[full-random-tall]
 decomposition/tests/test_pca.py::test_pca_explained_variance_empirical[full-correlated-tall]
 decomposition/tests/test_pca.py::test_pca_explained_variance_empirical[full-random-wide]
--------- generated xml file: /home/vsts/work/tmp_folder/test-data.xml ---------
=========================== short test summary info ============================
ERROR gw0
================== 1 skipped, 156 warnings, 1 error in 38.74s ==================
```

I don't remember what is the workaround for this...

ogrisel committed:
Another try to workaround pytest xdist bug

ogrisel committed:
Use the new svd_flip config consistently and update docstrings accordingly

ogrisel committed:
One more missing sign update in the FeatureUnion doctest

ogrisel committed:
Document the changed component sign heuristic

ogrisel committed:
More svd_flip update

ogrisel committed:
Attempt at fixing indentation problem for svd_solver docstring

ogrisel commented:
The Array API compliance tests pass for the new solver both for PyTorch with MPS on Apple M1 and for PyTorch cuda and CuPy with an NVIDIA GPU!

Will run some quick benchmarks on cuda and post them in the description.

ogrisel committed:
One more missing sign update in the FeatureUnion doctest

ogrisel committed:
Avoid SparsePCA's internal DictionaryLearning instance generating code as a pandas dataframe

ogrisel committed:
TST test fit_transform and transform equivalence

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel committed:
TST cleaner way to define X_train and X_test

ogrisel committed:
Add new benchmark script to check the robustness of the auto policy

ogrisel committed:
Ignore CSV file with collected benchmark data

ogrisel committed:
Adjust svd_solver='auto'

ogrisel committed:
Make test_pca_solver_equivalence more generic at dealing with unstable components

ogrisel committed:
Fix sklearn/decomposition/tests/test_incremental_pca.py:test_whitening

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel committed:
Update the docstring to reflect the new auto policy

ogrisel committed:
More explicit figure titles

ogrisel committed:
Update the docstring to mention the difference in numerical stability

ogrisel commented:
Hum, there is still a failure in the incremental PCA test that I could not observe locally. I will investigate next week. The rest of the PR can still be reviewed, though :)

ogrisel committed:
Improve solver equivalence tests

ogrisel committed:
DOC make compose doc more interesting and robust to rounding errors

ogrisel committed:
Mark the stricter test_pca_solver_equivalence with the arpack solver xfail for old scipy versions

ogrisel committed:
Make PCA(whiten=True).transform robust to rank deficient training data

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel commented on sklearn/decomposition/_dict_learning.py:
```diff
@@ -1227,7 +1227,7 @@ def dict_learning(
         positive_dict=positive_dict,
         transform_max_iter=method_max_iter,
-    )
+    ).set_output(transform="default")
```
This avoids calling svd_flip on a pandas dataframe in the subsequent call to `fit_transform`.

glemaitre replied:
Now I am thinking that since all our function are wrapping an estimator, we should most probably use this pattern everywhere. Anyway not link to the current PR.

ogrisel committed:
Update inline comment about clipping negative eigenvalues

ogrisel committed:
Simpler way to clip the whitening scale

ogrisel committed:
Test solver equivalence with float32 data

ogrisel commented on doc/whats_new/v1.4.rst:
```diff
@@ -208,12 +215,26 @@ Changelog
+  `whiten=True` and a value for `n_components` that is larger than the rank of
+  the training set, no longer returns infinite values when transforming
+  held-out data.
+  :pr:`27491` by :user:`Olivier Grisel <ogrisel>`.
```
I found this bug by refactoring `test_pca_solver_equivalence` to make it much stronger.

lorentzenchr replied:
Could it be factored out as separate PR?

ogrisel committed:
Forgot to update on condition based on variance_threshold

ogrisel commented on sklearn/decomposition/_pca.py:
```diff
@@ -175,26 +175,42 @@ class PCA(_BasePCA):
+            typically using LAPACK and select the components by postprocessing.
+            This solver is very efficient when the number of features is small
+            and not tractable otherwise (large memory footprint required to
+            materialize the covariance matrix). Also not that compared to the
```
```suggestion
            materialize the covariance matrix). Also note that compared to the
```

ogrisel committed:
Typo

ogrisel committed:
[azure parallel] [all random seeds]
test_pca_solver_equivalence
test_whitening

ogrisel committed:
Relax float32 test tolerance [azure parallel] [all random seeds]
test_pca_solver_equivalence
test_whitening

ogrisel commented:
Everything is green finally.

ogrisel committed:
More explicit conditions for assertions in test_pca_solver_equivalence [all random seeds] [azure parallel]

ogrisel committed:
[azure parallel] [all random seeds] test_pca_solver_equivalence

ogrisel commented:
After profiling with `py-spy`, I realize that the allocation for the centering for the computation of the mean is actually far for from negligible (around 40% of the time) and wastes a lot of memory since the goal is to compute a much smaller covariance matrix. We can probably fuse the operations (at least to some extent) to save memory and improve speed even further.

ogrisel committed:
ENH do not center a priori to spare memory and make the new solver run even faster

ogrisel commented:
Putting this back to draft mode: I have to re-run the benchmarks and check that the auto heuristic is still valid with the no-copy variant.

I also have to re-run the Array API tests on a cuda machine.

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel committed:
[azure parallel] [all random seeds]
test_pca_solver_equivalence"
test_whitening
test_pca_dtype_preservation

ogrisel committed:
[azure parallel] [all random seeds]
test_pca_solver_equivalence
test_whitening
test_pca_dtype_preservation

ogrisel commented:
The benchmarks look even better. We are now twice as fast as cuML for `PCA.fit` on tall and skinny data inputs.

Running `PCA.fit` on a laptop CPU on for millions of datapoints with a hundred features is now also a typically sub second task.

However, the improvement for `PCA.fit_transform` is not as strong (but still significant) when `n_components` is getting close to `n_features` as can be observed in https://github.com/scikit-learn/scikit-learn/pull/27491#issuecomment-1741079946.

Still this PR is very memory efficient for such datasets, so this is an additional side-benefit.

I also ran another round of `py-spy` profiling (for the `fit` part alone) and now `X.T @ X` and `xp.mean(X, axis=0)` are the two dominating operations on tall and skinny data, as expected. Note that they are memory bound, so attempting to multithread them is probably worthless. One could probably fuse them into a single kernel (using Cython, numba or triton) but that would require careful tiling/blocking and use of SIMD, so I think it's not worth it for now (and the expected speed-up would be around of factor 2 if all goes well).

When profiling `fit_transform`, the `transform` part is now dominating. Here again, fusing the centering into the GEMM kernel could reduce the memory bandwidth bottleneck and provide around a factor of 2 speed up, but again that would be complex so I don't think it's worth investing time in this.

ogrisel committed:
Propagate xp namespace to private methods

ogrisel committed:
Scale the covariance matrix before calling eigh to improve numerical stability

ogrisel committed:
Make it explicit when the X passed between private methods has been centered or not

ogrisel committed:
[azure parallel] [all random seeds]

test_pca_solver_equivalence
test_whitening
test_pca_dtype_preservation

ogrisel committed:
Improve test [azure parallel] [all random seeds] test_pca_solver_equivalence

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel committed:
OPTIM Ensure that components_ is contiguous

ogrisel commented:
I did another round of profiling and was surprised than the `transform` method for the new solver was an order of magnitude slower than similar sized operations in `_fit_full`. I realized that this was caused by the fact that the stored `components_` attribute was not contiguous (it was a 2D slice).

This is now fixed in 56a7793 and now makes `fit_transform(X_train)` much faster for the new solver, and `transform(X_test)` also much faster, both for the new solver and the `"full"` solver.

I will update the benchmark results in the description with the new state of this PR as well was the plots in https://github.com/scikit-learn/scikit-learn/pull/27491#issuecomment-1741079946 that are used to check that the "auto" policy works as intended.

- [ ] update benchmarks with numpy,
- [ ] update benchmarks with PyTorch and cuda,
- [ ] update benchmarks with PyTorch and MPS,
- [ ] update benchmarks used for the "auto" policy.

ogrisel committed:
Update auto policy and bench_pca_solvers.py

ogrisel commented:
I updated the "auto" policy in the previous commit. Here are the resulting plots from re-running `benchmarks/bench_pca_solvers.py` with the updated code:

- `fit` with fixed `n_features` and varying `n_samples`:

![image](https://github.com/scikit-learn/scikit-learn/assets/89061/93779700-4fdd-4ae6-b427-cd58bcc95ccd)

- `fit_transform` with fixed `n_features` and varying `n_samples`:

![image](https://github.com/scikit-learn/scikit-learn/assets/89061/0291bad4-0fbc-4c68-877a-a3434cdff544)

- `fit` with fixed `n_samples` and varying `n_features`:

![image](https://github.com/scikit-learn/scikit-learn/assets/89061/ce9076cb-85ea-4be5-9aaa-c0a420b4266e)

- `fit_transform` with fixed `n_samples` and varying `n_features`:

![image](https://github.com/scikit-learn/scikit-learn/assets/89061/2dac3941-2439-400f-bdfa-f865e221f7a9)


So with the updated code, the updated "auto" policy is near optimal both for `fit` and `fit_transform`.

Furthermore, the new method is an clearer performance improvement, in particular for `fit_transform` when `n_samples >> n_features` and `n_features` small enough.

ogrisel commented:
Ok, I have updated the benchmarks everywhere and the updated policy is near optimal.

In summary, the new solver gets a typical 10x speed-up over the previous default solver when `n_samples >> n_features`.

Furthermore, the new solver combined with PyTorch with the cuda device (via Array API) is 5x faster than NVIDIA's cuML on the same NVIDIA GPU (a V100 in this case).

ogrisel commented:
Note to reviewers, since to add this new solver I had to fix and improve other related things in the `sklearn.decomposition` module, I can attempt to split this PR into sub PRs (e.g. one per entry in the changelog: https://github.com/scikit-learn/scikit-learn/pull/27491/files#diff-0873f464b088aef8a6f2c9b516dadfb368c95df8e07d3b6edfe17a5777f00859).

lorentzenchr commented on sklearn/decomposition/_pca.py:
```diff
@@ -533,33 +559,67 @@ def _fit_full(self, X, n_components):
+
         else:
-            U, S, Vt = xp.linalg.svd(X, full_matrices=False)
+            assert self._fit_svd_solver == "covariance_eigh"
```
No `assert` in production code, only tests. This should be covered by ruff, did we turn it off?

ogrisel replied:
Personally I like asserts: it communicates better the assumption we make in the code and make it more legible. But I can turn the assertion into an inline comment to communicate the intended purpose of the `else:` branch of the condition.

However, I would like to avoid the situation where we do:

```python
if self._fit_svd_solver in ("A", "B"):
    do something
elif self._fit_svd_solver == "C":
    do something else
else:
     raise ValueError(f"Invalid solver: {self._fit_svd_solver}")
```

because the `ValueError` branch can never be reached because it's already validated.

At the same time if we refactor other branches of the condition, inline comments run the risk of not being kept up-to-date, while assert statements would.

lorentzenchr replied:
Good point. In that case leave it as is. Maybe match when in python 3.10 will help with those in the future.

glemaitre replied:
Adding a message with the assertion could be nicer.

lorentzenchr commented on sklearn/decomposition/_pca.py:
```diff
@@ -533,33 +559,67 @@ def _fit_full(self, X, n_components):
+            # of X. Note that the mean_ attribute is still needed to center
+            # test data in the transform method.
+            x_is_centered = False
+            C = X.T @ X
```
Why not use np.cov? At least a comment.

lorentzenchr commented on sklearn/decomposition/_pca.py:
```diff
@@ -656,8 +728,8 @@ def _fit_truncated(self, X, n_components, svd_solver):
-        X **= 2
-        total_var = xp.sum(xp.sum(X, axis=0) / N)
+        X_centered **= 2
+        total_var = xp.sum(xp.sum(X_centered, axis=0) / N)
```
```suggestion
        total_var = xp.sum(X_centered) / N
```

lorentzenchr reviewed (commented):
An early and fast skim of a review.

ogrisel committed:
Single call to xp.sum to compute total_var.

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

betatim commented on sklearn/decomposition/_base.py:
```diff
@@ -136,16 +136,32 @@ def transform(self, X):
-            X_transformed /= xp.sqrt(self.explained_variance_)
+            # For some solvers (such as "arpack" and "covariance_eigh"), on
+            # rank deficient data, some components can have a variance
+            # arbitrarily to zero, leading to non-finite results when
```
```suggestion
            # arbitrarily close to zero, leading to non-finite results when
```
is this the missing word?

betatim commented on sklearn/decomposition/_pca.py:
```diff
@@ -175,26 +175,42 @@ class PCA(_BasePCA):
-        If full :
-            run exact full SVD calling the standard LAPACK solver via
+            `n_components`: if the input data has fewer than 1000 features and
+            more than 10 times as many samples, then the more "covariance_eigh"
```
```suggestion
            more than 10 times as many samples, then the "covariance_eigh"
```
spurious word? or missing word?

betatim commented on sklearn/decomposition/_pca.py:
```diff
@@ -175,26 +175,42 @@ class PCA(_BasePCA):
+            solver is used. Otherwise, if the input data is larger than 500x500
+            and the number of components to extract is lower than 80% of the
+            smallest dimension of the data, then the more efficient
+            'randomized' method is enabled. Otherwise the exact "full" SVD is
```
```suggestion
            "randomized" method is enabled. Otherwise the exact "full" SVD is
```

betatim reviewed (commented):
Some comments, mostly for my understanding.

Generally it looks good I think. I can make sense of the tests and the code changes, and assume the new solver makes sense (given the tests pass), but not an expert on the topic really.

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Tim Head <betatim@gmail.com>

ogrisel commented on sklearn/utils/extmath.py:
```diff
@@ -863,24 +865,27 @@ def svd_flip(u, v, u_based_decision=True):
 
     if u_based_decision:
         # columns of u, rows of v, or equivalently rows of u.T and v
+        device = getattr(u, "device", None)
```
There is already a function there but it does something slightly different but maybe it will work. Let me try.

ogrisel replied:
I updated the existing inline comment.

ogrisel commented on sklearn/decomposition/_pca.py:
```diff
@@ -489,8 +510,11 @@ def _fit(self, X):
+        # datasets.
         X = self._validate_data(
-            X, dtype=[xp.float64, xp.float32], ensure_2d=True, copy=self.copy
+            X, dtype=[xp.float64, xp.float32], ensure_2d=True, copy=False
```
I will add an inline comment.

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel committed:
Use the device() utility function instead of getattr

ogrisel committed:
Improve inline comment to explain copy=False

ogrisel committed:
Improve comment to explain why we do not rely on numpy.cov

ogrisel commented:
I think I addressed all the comments of the first batch of review, except @lorentzenchr request to extra a sub PR here: https://github.com/scikit-learn/scikit-learn/pull/27491/#discussion_r1347941972

Do you still want me to split it into sub-PRs now that @betatim has reviewed this PR or is it fine to review everything at once here?

I can extract one sub-PR per item in the changelog if your prefer. But it might be a bit complex to do that in the correct order.

glemaitre commented on sklearn/utils/extmath.py:
```diff
@@ -843,12 +843,14 @@ def svd_flip(u, v, u_based_decision=True):
         Parameters u and v are the output of `linalg.svd` or
         :func:`~sklearn.utils.extmath.randomized_svd`, with matching inner
         dimensions so one can compute `np.dot(u * s, v)`.
+        u can be None if u_based_decision is False.
```
```suggestion
        u can be None if `u_based_decision` is False.
```

glemaitre commented on sklearn/utils/extmath.py:
```diff
@@ -843,12 +843,14 @@ def svd_flip(u, v, u_based_decision=True):
         :func:`~sklearn.utils.extmath.randomized_svd`, with matching inner
         dimensions so one can compute `np.dot(u * s, v)`. The input v should
         really be called vt to be consistent with scipy's output.
+        v can be None if u_based_decision is True.
```
```suggestion
        v can be None if `u_based_decision` is True.
```

glemaitre commented on doc/whats_new/v1.4.rst:
```diff
@@ -35,6 +35,13 @@ random sampling procedures.
       solvers (when fit on the same data again). The amount of change depends on the
       specified `tol`, for small values you will get more precise results.
 
+- |Enhancement| :class:`decomposition.PCA`, :class:`decomposition.SparsePCA` and
```
I am wondering if it has an impact on what was reported in https://github.com/scikit-learn/scikit-learn/issues/27620 when using MKL.

glemaitre commented on sklearn/decomposition/_pca.py:
```diff
@@ -460,17 +478,20 @@ def fit_transform(self, X, y=None):
         """
-        U, S, Vt = self._fit(X)
-        U = U[:, : self.n_components_]
+        U, S, Vt, X, x_is_centered, xp = self._fit(X)
```
```suggestion
        U, S, _, X, x_is_centered, xp = self._fit(X)
```

glemaitre commented on sklearn/decomposition/_pca.py:
```diff
@@ -505,8 +532,12 @@ def _fit(self, X):
         self._fit_svd_solver = self.svd_solver
         if self._fit_svd_solver == "auto":
+            # Tall and skinny problems are best handled by precomputing the
+            # covariance matrix.
+            if X.shape[1] <= 1000 and X.shape[0] >= 10 * X.shape[1]:
```
```suggestion
            # covariance matrix.
            if X.shape[1] <= 1_000 and X.shape[0] >= 10 * X.shape[1]:
```

glemaitre commented on sklearn/decomposition/_pca.py:
```diff
@@ -533,33 +562,77 @@ def _fit_full(self, X, n_components):
-            U, S, Vt = xp.linalg.svd(X, full_matrices=False)
+            assert self._fit_svd_solver == "covariance_eigh"
+            # In the following, we center the covariance matrix C a-posteriori
+            # (without centering the data X first) to avoid an unecessary copy
```
```suggestion
            # (without centering the data X first) to avoid an unnecessary copy
```

glemaitre commented on sklearn/decomposition/_pca.py:
```diff
@@ -533,33 +562,77 @@ def _fit_full(self, X, n_components):
+            # Besides, using `numpy.cov`, as of numpy 1.26.0, would not be
+            # memory efficient for our use case when `n_samples >> n_features`:
+            # `numpy.cov` centers a copy of the data before computing the
+            # matrix product instead of substracting a small `(n_features,
```
```suggestion
            # matrix product instead of subtracting a small `(n_features,
```

glemaitre commented on sklearn/decomposition/_pca.py:
```diff
@@ -533,33 +562,77 @@ def _fit_full(self, X, n_components):
+                * xp.reshape(self.mean_, (1, -1))
+            )
+            C /= n_samples - 1
+            eigenvals, Eigenvecs = xp.linalg.eigh(C)
```
Is there a reason for capital letter of `Eigenvecs`?

ogrisel replied:
It's a 2D matrix but yeah you are right, this is weird.

glemaitre commented on sklearn/decomposition/_kernel_pca.py:
```diff
@@ -363,9 +363,7 @@ def _fit_transform(self, K):
-        self.eigenvectors_, _ = svd_flip(
-            self.eigenvectors_, np.zeros_like(self.eigenvectors_).T
-        )
+        self.eigenvectors_, _ = svd_flip(self.eigenvectors_, None)
```
```suggestion
        self.eigenvectors_, _ = svd_flip(u=self.eigenvectors_, v=None)
```

glemaitre commented on sklearn/decomposition/tests/test_sparse_pca.py:
```diff
@@ -114,7 +115,10 @@ def test_initialization():
-    assert_allclose(model.components_, V_init / np.linalg.norm(V_init, axis=1)[:, None])
+
+    expected_components = V_init / np.linalg.norm(V_init, axis=1, keepdims=True)
+    expected_components = svd_flip(expected_components.T, None)[0].T
```
```suggestion
    expected_components = svd_flip(u=expected_components.T, v=None)[0].T
```

glemaitre reviewed (commented):
I need to look at the PCA test but I put a couple of comments (just nitpicks)

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel commented on sklearn/decomposition/_pca.py:
```diff
@@ -533,33 +562,77 @@ def _fit_full(self, X, n_components):
+                * xp.reshape(self.mean_, (1, -1))
+            )
+            C /= n_samples - 1
+            eigenvals, Eigenvecs = xp.linalg.eigh(C)
```
```suggestion
            eigenvals, eigenvecs = xp.linalg.eigh(C)
```

ogrisel commented on sklearn/decomposition/_pca.py:
```diff
@@ -533,33 +562,77 @@ def _fit_full(self, X, n_components):
+            C /= n_samples - 1
+            eigenvals, Eigenvecs = xp.linalg.eigh(C)
+            eigenvals = xp.flip(eigenvals, axis=0)
+            Eigenvecs = xp.flip(Eigenvecs, axis=1)
+
+            # The covariance matrix C is positive semi-definite by
+            # construction. However, the eigenvalues returned by xp.linalg.eigh
+            # can be slightly negative due to numerical errors. This would be
+            # an issue for the subsequent sqrt, hence the manual clipping.
+            eigenvals[eigenvals < 0.0] = 0.0
+            explained_variance_ = eigenvals
+
+            # Re-construct synthetic SVD components to be consistent with the
+            # other solvers.
+            S = xp.sqrt(eigenvals * (n_samples - 1))
+            Vt = Eigenvecs.T
```
```suggestion
            Eigenvecs = xp.flip(Eigenvecs, axis=1)

            # The covariance matrix C is positive semi-definite by
            # construction. However, the eigenvalues returned by xp.linalg.eigh
            # can be slightly negative due to numerical errors. This would be
            # an issue for the subsequent sqrt, hence the manual clipping.
            eigenvals[eigenvals < 0.0] = 0.0
            explained_variance_ = eigenvals

            # Re-construct synthetic SVD components to be consistent with the
            # other solvers.
            S = xp.sqrt(eigenvals * (n_samples - 1))
            Vt = eigenvecs.T
```

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

ogrisel committed:
Attempt at merging main with sparse input support for PCA

ogrisel commented:
Hi @ivirshup! I have tried to merge the conflicts that appeared after merging your PR to add sparse input data support to PCA but this PR breaks your tests (in particular for the `transform` method).

I had a quick look at the test and the diff of the PR but I don't understand what's wrong. Would you mind having a look?

ogrisel committed:
Fix typo

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel committed:
Fix broken tests

ogrisel committed:
Increase rtol for array api compliance test with float32 pytorch

ogrisel committed:
Accept sparse input data for svd_solver='covariance_eigh'

ogrisel commented:
I found the cause of the failing test. Since it was easy to add sparse data support for this solver (as it centers the covariance matrix a posteriori) I just did it as part of this PR.

ogrisel committed:
Fix expected error message in test

ivirshup commented:
Nice! Sorry, traveling to a conference over the weekend.

How is the performance? I played around with this a bit last week and it seemed like it should be quite good for sparse input.

ogrisel commented:
I have not benchmarked sparse inputs but I expect a similar scalability profile vs ARPACK that on the dense data results posted in the comments above.

ogrisel commented:
Once we add sparse support for the randomized solver (in a separate PR), I propose to adjust the `svd_solver="auto"` strategy to also handle the sparse input case to automatically switch between randomized, covariance_eigh and arpack based on similar benchmarks. But let's keep that for later and keep the explicit solver selection for sparse data for now.

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel committed:
DOC move changelog entries to target 1.5

ogrisel committed:
DOC cleanup left over entry that was meant to be moved to 1.5

ogrisel committed:
Fix bad conflict resolution

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel commented:
I updated the PR to sync it with `main`, resolve conflicts and target 1.5 instead o 1.4.

glemaitre commented on sklearn/decomposition/_base.py:
```diff
@@ -138,21 +136,37 @@ def transform(self, X):
-            X, accept_sparse=("csr", "csc"), dtype=[xp.float64, xp.float32], reset=False
+            X, dtype=[xp.float64, xp.float32], accept_sparse=("csr", "csc"), reset=False
+        )
+        return self._transform(
+            X,
+            xp=xp,
+            x_is_centered=False,
         )
```
```suggestion
        return self._transform(X, xp=xp, x_is_centered=False)
```

glemaitre reviewed (commented):
It looks good. I'll just have a IRL question regarding the centering I think.

ogrisel commented on sklearn/decomposition/_pca.py:
```diff
@@ -591,22 +676,32 @@ def _fit_full(self, X, n_components):
+        # - release the memory used by the discarded components,
+        # - ensure that the kept components are allocated contiguously in memory
+        #   to make the transform method faster by leveraging cache locality.
+        self.components_ = xp.asarray(components_[:n_components, :], copy=True)
```
No it's not.

lorentzenchr commented:
@glemaitre @ogrisel Do you bring this over the finish line or do you prefer the 2nd review approval from me?

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel commented on sklearn/decomposition/_pca.py:
```diff
@@ -448,41 +467,50 @@ def fit_transform(self, X, y=None):
 
-        return U
+            return U
+        else:
```
For now this is the case, but we might want to also do that for the `randomized` solver because of the numerical discrepancy reported in #18941.

But let's do that in a dedicated follow-up PR.

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

ogrisel committed:
Use assert_allclose in test_whitening

ogrisel committed:
Move changed models entry to the existing section of the change log

ogrisel commented on sklearn/decomposition/_pca.py:
```diff
@@ -448,41 +467,50 @@ def fit_transform(self, X, y=None):
-        U, S, Vt = self._fit(X)
-        U = U[:, : self.n_components_]
+        U, S, _, X, x_is_centered, xp = self._fit(X)
+        if U is not None:
```
Resolved by https://github.com/scikit-learn/scikit-learn/pull/27491#discussion_r1452785739.

ogrisel commented:
@lorentzenchr I updated the PR to take the latest batch of review comments into account.

ogrisel commented:
I re-ran the benchmark script and the auto strategy is still near-optimally picking the fastest solver.

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel commented:
BTW, it would be interesting to try this new solver with dask in an out-of-core setting where `n_samples` is very large when #28588 is merged.

lorentzenchr commented on doc/whats_new/v1.5.rst:
```diff
@@ -31,6 +31,13 @@ Changed models
 
+- |Enhancement| :class:`decomposition.PCA`, :class:`decomposition.SparsePCA` and
+  :class:`decomposition.TruncatedSVD` now set the sign of the `components_`
+  attribute based on the components values instead of using the transformed
```
```suggestion
  attribute based on the component values instead of using the transformed
```

lorentzenchr commented on doc/whats_new/v1.5.rst:
```diff
@@ -165,10 +172,33 @@ Changelog
+- |Efficiency| :class:`decomposition.PCA` with `svd_solver="full"` now assigns
+  a contiguous `components_` attribute instead of an non-contiguous slice of
+  the singular vectors. When `n_components << n_features`, this can save some
+  memory and more importantly help speed-up subsequent calls to the `transform`
```
```suggestion
  memory and, more importantly, help speed-up subsequent calls to the `transform`
```

lorentzenchr commented on doc/whats_new/v1.5.rst:
```diff
@@ -165,10 +172,33 @@ Changelog
+- |Enhancement| :class:`decomposition.PCA` now supports a new solver option
+  named `svd_solver="covariance_eigh"` which offers an order of magnitude
+  speed-up and reduced memory usage for datasets with a large number of data
+  points and a small number of features (say, less than 1000). The
```
```suggestion
  points and a small number of features (say, `n_samples >> 1000 > n_features`). The
```

lorentzenchr commented on doc/whats_new/v1.5.rst:
```diff
@@ -165,10 +172,33 @@ Changelog
+- |Fix| :class:`decomposition.PCA` fit with `svd_solver="arpack"`,
+  `whiten=True` and a value for `n_components` that is larger than the rank of
+  the training set, no longer returns infinite values when transforming
+  held-out data.
```
```suggestion
  hold-out data.
```

lorentzenchr reviewed (commented):
Review of non-code.

lorentzenchr commented on sklearn/decomposition/_pca.py:
```diff
@@ -177,26 +180,42 @@ class PCA(_BasePCA):
+            Precompute the covariance matrix (on centered data), run a
+            classical eigenvalue decomposition on the covariance matrix
+            typically using LAPACK and select the components by postprocessing.
+            This solver is very efficient when the number of features is small
+            and not tractable otherwise (large memory footprint required to
+            materialize the covariance matrix). Also note that compared to the
```
```suggestion
            This solver is very efficient for n_samples >> n_features and small n_features.
            It is, however, not tractable otherwise for large n_features (large memory footprint required to
            materialize the covariance matrix). Also note that compared to the
```
Do we know the computation or memory complexity in terms of n_features, like memory is `O(n_features**2)` and computation `O(n_features**3)`?

ogrisel replied:
I know that it's very bad empirically (see the benchmarks above). The memory is definitely `O(n_features**2)`. For computation I am not 100% sure but indeed it should be dominated by the `eigh` call on a `n_features` square matrix, so it should indeed be cubic w.r.t. `n_features`.

lorentzenchr commented on sklearn/decomposition/_pca.py:
```diff
@@ -449,39 +470,49 @@ def fit_transform(self, X, y=None):
                 "PCA with svd_solver='arpack' is not supported for Array API inputs."
             )
 
+        # Validate the data, without ever forcing a copy as any solvers that
+        # support sparse input data and the `covariance_eigh` solver are
+        # written in a way to avoid the need for any inplace modification of
```
```suggestion
        # Validate the data, without ever forcing a copy as any solver that
        # supports sparse input data and the `covariance_eigh` solver are
        # written in a way to avoid the need for any inplace modification of
```

lorentzenchr commented on sklearn/decomposition/_pca.py:
```diff
@@ -449,39 +470,49 @@ def fit_transform(self, X, y=None):
+        # Validate the data, without ever forcing a copy as any solvers that
+        # support sparse input data and the `covariance_eigh` solver are
+        # written in a way to avoid the need for any inplace modification of
+        # the input data contrary to the other solvers. Forcing a copy here
+        # would be wasteful when using large datasets. The copy will happen
```
```suggestion
        # the input data contrary to the other solvers.
        # The copy will happen
```

lorentzenchr commented on sklearn/decomposition/_pca.py:
```diff
@@ -523,33 +557,88 @@ def _fit_full(self, X, n_components):
         else:
-            U, S, Vt = xp.linalg.svd(X, full_matrices=False)
+            assert self._fit_svd_solver == "covariance_eigh"
+            # In the following, we center the covariance matrix C a-posteriori
```
```suggestion
            # In the following, we center the covariance matrix C afterwards
```

lorentzenchr commented on sklearn/decomposition/_pca.py:
```diff
@@ -523,33 +557,88 @@ def _fit_full(self, X, n_components):
+            # memory efficient for our use case when `n_samples >> n_features`:
+            # `numpy.cov` centers a copy of the data before computing the
+            # matrix product instead of subtracting a small `(n_features,
+            # n_features)` square matrix, a posteriori, as we do below.
```
```suggestion
            # n_features)` square matrix from the gram matrix X.T @ X, as we do below.
```
Note that this is numerically more unstable than np.cov.

lorentzenchr commented on sklearn/decomposition/_pca.py:
```diff
@@ -523,33 +557,88 @@ def _fit_full(self, X, n_components):
+            eigenvals, eigenvecs = xp.linalg.eigh(C)
+
+            # The following can happen when X is a scipy sparse matrix.
+            if isinstance(eigenvals, np.matrix):
```
I think we should not use `np.matrix` anywhere.

ogrisel replied:
The point of this line is precisely to get rid of such `np.matrix` occurrences that typically show up when using accepting `scipy.sparse.csr_matrix` and the like as inputs. `scipy.sparse.csr_array` does not lead to this problem anymore but we need to be backward compatible with old scipy for some time still before being able to deprecate support for `csr_matrix`.

ogrisel replied:
```python
>>> import numpy as np
>>> from scipy.sparse import rand
>>> X = rand(100_000, 5)
>>> X
<100000x5 sparse matrix of type '<class 'numpy.float64'>'
        with 5000 stored elements in COOrdinate format>
>>> C = X.T @ X
>>> C
<5x5 sparse matrix of type '<class 'numpy.float64'>'
        with 25 stored elements in Compressed Sparse Row format>
>>> n_samples = X.shape[0]
>>> X_mean = X.mean(axis=0)
>>> X_mean
matrix([[0.00531109, 0.00497345, 0.00498098, 0.00488378, 0.00509504]])
>>> C -= (n_samples * X_mean.reshape(-1, 1) * X_mean.reshape(1, -1))
>>> C
matrix([[ 3.59159095e+02, -1.48415344e+00,  2.06023127e+00,
         -2.50133851e-02, -2.62356078e+00],
        [-1.48415344e+00,  3.33571655e+02,  7.52427579e-01,
          2.02376720e+00,  6.44612078e-01],
        [ 2.06023127e+00,  7.52427579e-01,  3.27488715e+02,
          2.18887774e+00, -8.94077160e-01],
        [-2.50133851e-02,  2.02376720e+00,  2.18887774e+00,
          3.24096148e+02,  8.62317950e-01],
        [-2.62356078e+00,  6.44612078e-01, -8.94077160e-01,
          8.62317950e-01,  3.37911475e+02]])
>>> C /= n_samples - 1
>>> eigenvals, eigenvecs = np.linalg.eigh(C)
>>> eigenvals
array([0.00322734, 0.00327965, 0.00334087, 0.00337746, 0.00359712])
>>> eigenvecs
matrix([[-0.01358159,  0.06717155, -0.023717  ,  0.13244301, -0.98853421],
        [ 0.13346428,  0.27465896, -0.94081235,  0.13532164,  0.05753191],
        [ 0.41103831, -0.88445837, -0.2085449 , -0.03245091, -0.06509116],
        [-0.89918035, -0.36812082, -0.22385077,  0.07644526,  0.00295262],
        [ 0.0672684 , -0.04765382,  0.14390791,  0.97839152,  0.12346913]])
```

lorentzenchr replied:
How about calling `eigenvals = xp.asarray(eigenvals)` and a TODO note for sparse matrix cleaning (once that happens)?

lorentzenchr commented on sklearn/decomposition/_pca.py:
```diff
@@ -523,33 +557,88 @@ def _fit_full(self, X, n_components):
+            eigenvals[eigenvals < 0.0] = 0.0
+            explained_variance_ = eigenvals
+
+            # Re-construct synthetic SVD components to be consistent with the
```
```suggestion
            # Re-construct SVD of centered X indirectly and make it consistent with the
```

ogrisel commented on sklearn/decomposition/_base.py:
```diff
@@ -138,21 +136,33 @@ def transform(self, X):
+            # copy of X or mutating the data passed by the caller.
+            # This also works for sparse X without having to wrap it into a
+            # linear operator a priori.
+            X_transformed = X @ self.components_.T
```
Interesting, that might make this method even more competitive for the `fit_transform` case. I would need to re-run the benchmarks...

lorentzenchr replied:
At least a review that was worthwhile.

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

ogrisel committed:
Merge branch 'main' into pca-covariance-eigh

ogrisel committed:
Wrap paragraphs in multiline comments

lorentzenchr commented on sklearn/decomposition/tests/test_incremental_pca.py:
```diff
@@ -384,25 +384,38 @@ def test_singular_values():
+        # Since the data is rank deficient, some components are pure noise. We
+        # should not expect those dimensions to carry any signal and their
+        # values might be arbitrarily changed by implementation details of the
+        # internal SVD solver. We therefore mask them out before comparison.
```
```suggestion
        # internal SVD solver. We therefore filter them out before comparison.
```

lorentzenchr commented on sklearn/decomposition/tests/test_incremental_pca.py:
```diff
@@ -384,25 +384,38 @@ def test_singular_values():
         pca = PCA(whiten=True, n_components=nc).fit(X)
         ipca = IncrementalPCA(whiten=True, n_components=nc, batch_size=250).fit(X)
 
+        # Since the data is rank deficient, some components are pure noise. We
+        # should not expect those dimensions to carry any signal and their
+        # values might be arbitrarily changed by implementation details of the
+        # internal SVD solver. We therefore mask them out before comparison.
```
Do we want to test rank deficient data or could we simplify the test with other data X?

ogrisel replied:
I think the test is stronger this way. `IncrementalPCA` is typically used when `n_samples >> n_features` and with the default value of `n_components=None` this case is bound to happen in practice.

ogrisel replied:
Note that `test_pca_solver_equivalence` covers even more cases and actually found a bug in for a related case with `PCA(whiten=True, svd_solver="arpack")` (the one documented in the changelog), so it's actually useful to (also) test with rank deficient data.

lorentzenchr reviewed (approved):
LGTM (despite the fact that I'm not that much into PCA 😏)

lorentzenchr commented:
@ogrisel Ping when ready to merge.

ogrisel committed:
Update sklearn/decomposition/tests/test_incremental_pca.py

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

ogrisel committed:
Wrap paragraphs in changelog

ogrisel commented:
> @ogrisel Ping (me/us) when ready to merge.

@lorentzenchr this is ready from my point of view.

ogrisel commented:
I forgot about https://github.com/scikit-learn/scikit-learn/pull/27491#discussion_r1559721096. Let me give it a shot.

ogrisel committed:
Do not test on isinstance(..., np.matrix) and only use Array API for consistency

lorentzenchr commented on sklearn/decomposition/_pca.py:
```diff
@@ -522,33 +558,93 @@ def _fit_full(self, X, n_components):
-            # solver by default though (assuming both are built against the
-            # same BLAS).
-            U, S, Vt = linalg.svd(X, full_matrices=False)
+        if isinstance(self.mean_, np.matrix):
```
Same here, can we avoid np.matrix in code?

ogrisel replied:
Done.

lorentzenchr reviewed (commented):
Last review comment 🤞

ogrisel committed:
Do not test on isinstance(..., np.matrix) and only use Array API for consistency

ogrisel commented:
@lorentzenchr this should be good to go now. For information I also did the following experiment to check numerical stability in the `n_samples >> n_features > rank` setting:

```python
In [1]: import numpy as np

In [2]: X_low = np.random.randn(int(1e7), 5)

In [3]: W = np.random.randn(5, 30)

In [4]: X = X_low @ W

In [5]: X += np.random.randn(*X.shape) * 1e-12  # rank 5 up to eps-level noise

In [6]: from sklearn.decomposition import PCA

In [7]: PCA().fit(X).explained_variance_
Out[7]: 
array([5.13565104e+01, 4.16859476e+01, 2.90827456e+01, 2.34093563e+01,
       1.54211597e+01, 5.95719954e-14, 5.63161175e-14, 5.02011137e-14,
       4.21787206e-14, 2.90082256e-14, 2.88704010e-14, 1.78452632e-14,
       1.53384409e-14, 1.36259394e-14, 1.14775441e-14, 9.65281151e-15,
       4.52684746e-15, 2.73330102e-15, 2.23056558e-17, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00])

In [8]: PCA(svd_solver="covariance_eigh").fit(X).explained_variance_
Out[8]: 
array([5.13565104e+01, 4.16859476e+01, 2.90827456e+01, 2.34093563e+01,
       1.54211597e+01, 5.95719954e-14, 5.63161175e-14, 5.02011137e-14,
       4.21787206e-14, 2.90082256e-14, 2.88704010e-14, 1.78452632e-14,
       1.53384409e-14, 1.36259394e-14, 1.14775441e-14, 9.65281151e-15,
       4.52684746e-15, 2.73330102e-15, 2.23056558e-17, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00])

In [9]: PCA(svd_solver="full").fit(X).explained_variance_
Out[9]: 
array([5.13565104e+01, 4.16859476e+01, 2.90827456e+01, 2.34093563e+01,
       1.54211597e+01, 1.00568294e-24, 1.00282867e-24, 1.00269022e-24,
       1.00233412e-24, 1.00208860e-24, 1.00173990e-24, 1.00155317e-24,
       1.00116836e-24, 1.00078966e-24, 1.00051726e-24, 1.00049199e-24,
       1.00025070e-24, 1.00023886e-24, 1.00011534e-24, 9.99640288e-25,
       9.99398032e-25, 9.99344471e-25, 9.99024388e-25, 9.98812971e-25,
       9.98585628e-25, 9.98398337e-25, 9.98292080e-25, 9.97761657e-25,
       9.97621196e-25, 9.97159309e-25])
```

So I think the new solver is fine. The non-eps level components of the spectrum are recovered correctly and the other are below eps level for both `covariance_eigh` and `full` solvers.

Tialo commented on sklearn/decomposition/_pca.py:
```diff
@@ -176,26 +180,43 @@ class PCA(_BasePCA):
 
         .. versionadded:: 0.18.0
 
+        .. versionchanged:: 1.4
```
shouldn't there be 1.5?

lorentzenchr replied:
Yes. Well spotted. Do you want to open a PR to fix it?

Tialo replied:
Yes I'll do it

rutujagurav commented:
> BTW, it would be interesting to try this new solver with dask in an out-of-core setting where `n_samples` is very large when #28588 is merged.

Hello. Are there any plans to support a `partial_fit()` method with this solver to do exact PCA on datasets with a large number of samples but a modest number of features? Example: n_samples in the millions, n_features is 10k, so X.T @ X is small enough to hold in memory, but X itself isn't. Currently, the only available method is the SVD-based `IncrementalPCA()`.

betatim commented:
It is probably better to create a new issue for your question. It lets you add context, reference this PR and gives a space for discussion.

Rebooting very old PRs like this is not the right way.
